cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,246 @@
+// +---------------+-----------------------+------------------------|
+// | Node present: | Install control plane | Join the cluster as    |
+// |               | and be schedulable    | just a node            |
+// |---------------+-----------------------+------------------------|
+// | Node absent:  | Install control plane | Invalid configuration  |
+// |               | and be unscheduleable |                        |
+// +---------------+-----------------------+------------------------+
+type MachineRole string
+
+const (
+	MasterRole = "Master"
+	NodeRole   = "Node"
+)
+
+type MachineStatus struct {
+	// If the corresponding Node exists, this will point to its object.
+	// +optional
+	NodeRef *corev1.ObjectReference `json:"nodeRef,omitempty"`
+
+	// When was this status last observed
+	// +optional
+	LastUpdated metav1.Time `json:"lastUpdated,omitempty"`
+
+	// The current versions of software on the corresponding Node (if it
+	// exists). This is provided for a few reasons:
+	//
+	// 1) It is more convenient than checking the NodeRef, traversing it to
+	//    the Node, and finding the appropriate field in Node.Status.NodeInfo
+	//    (which uses different field names and formatting).
+	// 2) It removes some of the dependency on the structure of the Node,
+	//    so that if the structure of Node.Status.NodeInfo changes, only
+	//    machine controllers need to be updated, rather than every client
+	//    of the Machines API.
+	// 3) There is no other way simple way to check the ControlPlane
+	//    version. A client would have to connect directly to the apiserver
+	//    running on the target node in order to find out its version.
+	// +optional
+	Versions *MachineVersionInfo `json:"versions,omitempty"`
+
+	// In the event that there is a terminal problem reconciling the
+	// Machine, both ErrorReason and ErrorMessage will be set. ErrorReason
+	// will be populated with a succinct value suitable for machine
+	// interpretation, while ErrorMessage will contain a more verbose
+	// string suitable for logging and human consumption.
+	//
+	// These fields should not be set for transitive errors that a
+	// controller faces that are expected to be fixed automatically over
+	// time (like service outages), but instead indicate that something is
+	// fundamentally wrong with the Machine's spec or the configuration of
+	// the controller, and that manual intervention is required. Examples
+	// of terminal errors would be invalid combinations of settings in the
+	// spec, values that are unsupported by the controller, or the
+	// responsible controller itself being critically misconfigured.
+	//
+	// Any transient errors that occur during the reconcilation of Machines
+	// can be added as events to the Machine object and/or logged in the
+	// controller's output.
+	// +optional
+	ErrorReason *MachineStatusError `json:"errorReason,omitempty"`
+	// +optional
+	ErrorMessage *string `json:"errorMessage,omitempty"`
```

Comment:
Why using a pointer here instead of the value?

Reply:
To check if it maybe `nil`.

Reply:
Whats the benefit from using 
```golang
if machine.Status.ErrorMessage == nil {...}
```
over
```golang
if machine.Status.ErrorMessage == "" {...}
```
?

The same for `ErrorReason`. The potential error reasons are string constants.
Setting `ErrorReason` becomes annoying when its a pointer to a string.
Or am i missing something very obvious?

Reply:
It's just a convention for option string fields in Kubernetes API (there many examples in [`types.go`](https://github.com/kubernetes/kubernetes/blob/da9a4d5da9ecc6f687c873fb990d662b1f9f98b2/pkg/apis/core/types.go)).